crates/farming/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+
+        /// The farming pallet id, used for deriving pool accounts.
+        #[pallet::constant]
+        type FarmingPalletId: Get<PalletId>;
+
+        /// The treasury pallet id, used for deriving its sovereign account ID.
+        #[pallet::constant]
+        type TreasuryPalletId: Get<PalletId>;
+
+        /// Currency handler to transfer tokens
+        type MultiCurrency: MultiReservableCurrency<AccountIdOf<Self>, CurrencyId = CurrencyId>;
+
+        /// Reward pools to track stake
+        type RewardPools: RewardsApi<
+            CurrencyIdOf<Self>, // pool id is the lp token
+            AccountIdOf<Self>,
+            BalanceOf<Self>,
+            CurrencyId = CurrencyIdOf<Self>,
+        >;
+
+        /// Weight information for the extrinsics.
+        type WeightInfo: WeightInfo;
+    }
+
+    // The pallet's events
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        RewardScheduleUpdated {
+            pool_id: CurrencyIdOf<T>,
+            reward_schedule: RewardScheduleOf<T>,
+        },
+        RewardScheduleRemoved {
+            pool_id: CurrencyIdOf<T>,
+        },
+        RewardClaimed {
+            account_id: AccountIdOf<T>,
+            pool_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        },
+        RewardDistributed {
+            pool_id: CurrencyIdOf<T>,
+            currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        ScheduleNotFound,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+        fn on_initialize(n: T::BlockNumber) -> Weight {
+            if let Err(e) = Self::begin_block(n) {
+                sp_runtime::print(e);
+            }
+            Weight::from_ref_time(0 as u64)
+        }
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn reward_schedules)]
+    pub type RewardSchedules<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        CurrencyIdOf<T>, // lp token
+        Blake2_128Concat,
+        CurrencyIdOf<T>, // reward currency
+        RewardScheduleOf<T>,
+        OptionQuery,
+    >;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    // The pallet's dispatchable functions.
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Create or overwrite the reward schedule, if a reward schedule
+        /// already exists for the rewards currency it will first distribute
+        /// any remaining tokens to the rewards pool
+        #[pallet::weight(T::WeightInfo::update_reward_schedule())]
+        #[transactional]
+        pub fn update_reward_schedule(
+            origin: OriginFor<T>,
+            pool_id: CurrencyIdOf<T>,
+            currency_id: CurrencyIdOf<T>,
+            reward_schedule: RewardScheduleOf<T>,
+        ) -> DispatchResult {
```

Comment:
I'm not sure having the `RewardScheduleOf` argument is the best api. It works fine for inserting, but modifying existing schedules is problematic. You can't double the blockrewards, for example, without restarting the schedule. Part of the problem is that you can't really know in advance when the call will be executed, due to this having to go through governance..

Reply:
Please check the latest commit, this extrinsic now takes `per_period` and the `amount` as arguments.